### PR TITLE
feat: limit the maximum number of concurrent login attempts

### LINKED ISSUE #3423: Make it harder to brute force passwords
# Summary

Today, we immediately return the results of a login attempt. This allows attackers to quickly retry failed attempts in an effort to guess the password.

Also, today, the time taken to respond is different depending on whether a usernames exists or not, allowing the attacker to guess whether a userid is valid.

# Motivation

N/A

# Proposal

We should wait a fixed time, say one second, from the time at which the request was received, before returning the result. Waiting from the time at which the request was received, ensures that we don't leak any information due to different code paths being exercised in response to the validity of the username or password.

For extra credit, we may want to increase the delay after each consecutive unsuccessful login attempt

### PR DESCRIPTION
**Pull request by alexmt**
Closes #3423

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -114,8 +120,20 @@ var backoff = wait.Backoff{
 var (
 	clientConstraint = fmt.Sprintf(">= %s", common.MinClientVersion)
 	baseHRefRegex    = regexp.MustCompile(`<base href="(.*)">`)
+	// limits number of concurrent login requests to prevent password brute forcing. If set to 0 then no limit is enforced.
+	maxConcurrentLoginRequestsCount = 50
 )
 
+func init() {
+	if valStr := os.Getenv(maxConcurrentLoginRequestsCountEnv); valStr != "" {
```
Comment (jannfis) [inline server/server.go:128]: Hm, maybe we should move `parseNumFromEnv()` from `util/session` (https://github.com/argoproj/argo-cd/blob/7660e40fdcd1d870528f118a0618a69a8331426a/util/session/sessionmanager.go#L114) to a generic util function so we can reuse it whenever we need to parse a number from an environment variable.

File: server/session/ratelimiter.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+func (redis *redisStateStorage) set(key string, value interface{}, expiration time.Duration) error {
+	return redis.client.Set(key, value, expiration).Err()
+}
+
+func (redis *redisStateStorage) get(key string) (int, error) {
+	return redis.client.Get(key).Int()
+}
+
+func NewLoginRateLimiter(storage stateStorage, maxNumber int) func() (util.Closer, error) {
```
Comment (jannfis) [inline server/session/ratelimiter.go:53]: Can we have some comments please?

**Review by jannfis — CHANGES_REQUESTED** [review]
LGTM so far with minor comments. But we should aim for a higher test coverage IMHO, as this functionality will affect user experience.

File: server/session/ratelimiter.go (hunk 2 above)
Comment (alexmt) [reply server/session/ratelimiter.go:58]: done.

File: server/server.go (hunk 1 above)
Comment (alexmt) [reply server/server.go]: good suggestion! done.

File: server/session/ratelimiter.go (hunk 2 above)
Comment (jannfis) [reply server/session/ratelimiter.go:58]: :+1:

**Review by jannfis — APPROVED** [review]
Awesome, thanks. LGTM now.

**Pull request merged**